datashader/layout.py

```diff
@@ -0,0 +1,145 @@
+"""Assign coordinates to the nodes of a graph.
+
+Timothee Poisot's `nxfa2` is the original implementation of the main
+algorithm.
+
+.. _nxfa2:
+   https://github.com/tpoisot/nxfa2
+"""
+
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+import pandas as pd
+import param
+import scipy as sp
+
+
+def _convert_graph_with_positions_to_dataframes(graph, pos):
+    """
+    Convert NetworkX graph with associated positions into two dataframes.
+
+    In a NetworkX graph, each edge can have its own independent attributes. One
+    edge can have a different set of attributes than another edge. This means
+    we have to assign a default weight value when converting to dataframes.
+    """
+    nodes = pd.DataFrame()
+    for node, xy in zip(graph, pos):
+        nodes = nodes.append({'id': node, 'x': xy[0], 'y': xy[1]}, ignore_index=True)
+
+    nodes['id'].astype(np.int32)
+    nodes = nodes.set_index('id')
+
+    edges = pd.DataFrame()
+    for edge in graph.edges():
+        edge_attributes = graph[edge[0]][edge[1]]
+        if 'weight' in edge_attributes:
+            weight = edge_attributes['weight']
+        else:
+            weight = 1
+        edges = edges.append({'source': edge[0], 'target': edge[1], 'weight': weight}, ignore_index=True)
+
+    edges['source'].astype(np.int32)
+    edges['target'].astype(np.int32)
+
+    return nodes, edges
+
+
+class forceatlas2_layout(param.ParameterizedFunction):
+    """
+    Assign coordinates to the nodes of a graph.
+
+    This is a force-directed graph layout algorithm.
+
+    .. _ForceAtlas2:
+       http://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0098679&type=printable
+    """
+
+    def __call__(self, graph, iterations=10, linlog=False, pos=None, nohubs=False, k=None, dim=2):
+        """
+        Parameters
+        ----------
+        graph : networkx.Graph
+            The NetworkX graph to layout
+        iterations : int
+            Number of iterations
+        linlog : bool
+            Whether to use logarithmic attraction force
+        pos : ndarray
+            Initial positions for the given nodes
+        nohubs : bool
+            Whether to grant authorities (nodes with a high indegree) a
+            more central position than hubs (nodes with a high outdegree)
+        k : float
+            Compensates for the repulsion for nodes that are far away
+            from the center. Defaults to the inverse of the number of
+            nodes.
+        dim : int
+            Coordinate dimensions of each node.
+
```

Comment:
If this is a ParameterizedFunction, it should have parameters, i.e. declare ``iterations = param.Integer(10,doc="Number of iterations")`` at the class level, with ``__call__`` set up as other ParameterizedFunctions are (no explicit named arguments like that) and no docstring like this.